### ISSUE TITLE
Implement JsonSerializable in Phalcon\Mvc\Model, Phalcon\Mvc\Model\Resultset and Phalcon\Mvc\Model\Row

### ISSUE THREAD

**Issue by stibiumz** [phalcon/cphalcon#10273]
Once dropped PHP < 5.4 support, http://php.net/jsonserializable

**Comment by patrick-zippenfenig** [comment]
+1 usefull for APIs. I'd be happy to contribute

**Comment by patrick-zippenfenig** [comment]
Started implementing right now :)

**Comment by stibiumz** [comment]
@patrick-zippenfenig is then Phalcon 2.0 built for PHP >= 5.4? I think that it's supporting 5.3...

**Comment by patrick-zippenfenig** [comment]
Hm.. I always tend to forget version differences. I could just implement jsonSerialize without actually implementing the interface and hope php >5.3 just looks for jsonSerialize. Otherwise we could define the interface for php 5.3 (http://stackoverflow.com/questions/9683687/json-encode-is-not-working-with-array-of-objects/27809271#27809271) I'll give it a try

**Comment by patrick-zippenfenig** [comment]
PHP >5.3 does check for JsonSerializable interface and have to call jsonSerialize() manually: https://github.com/patrick-zippenfenig/cphalcon/commit/fbe06ad493f00b329d40c20e66d44a9a0ecee6e3
````
$robots = Robots::find();
echo json_encode( $robots->jsonSerialize() );
````
@andresgutierrez would this behaviour comply with phalcon guidelines?

**Comment by andresgutierrez** [comment]
I think the following would be better:

```php
$robots = Robots::find();
echo json_encode( $robots->toJson() );
```

**Comment by sergeyklay** [comment]
I agree with @andresgutierrez `toJson` is more readable than `jsonSerialize`. Also I thing that `toJson` is meets common CS like `toArray`.

**Comment by virgofx** [comment]
:+1: toJson()

**Comment by KorsaR-ZN** [comment]
@andresgutierrez, @sergeyklay, @virgofx
This requirement interface,
please see http://php.net/jsonserializable

**Comment by virgofx** [comment]
Ah well we could always alias toJson().

**Comment by patrick-zippenfenig** [comment]
PHP >5.4 calls ````jsonSerialize()```` automatically for objects implementing ````JsonSerializable```` interface:
````php
class Robots extends Mvc\Model {
  public function jsonSerialize() {
    return array('name' => $this->name);
  }
}
$robots = Robots::find();
echo json_encode( $robots );
````
Interface ````JsonSerializable````  is not available in PHP 5.3. It would be possible to define this interface manually (not sure if possible with zephir): 
````php
if (!interface_exists('JsonSerializable')) {
   interface JsonSerializable {
      public function jsonSerialize();
   }
}
````
PHP 5.3 user would have to call ````json_encode( $robots->jsonSerialize() );```` manually

**Comment by Green-Cat** [comment]
I would quite like that, it would greatly simplify one of my APIs.

**Comment by Green-Cat** [comment]
@sergeyklay Implemented in #10835
